File: trading-bot-moomoo/risk/manager.py
"""Risk manager: enforces position limits, daily loss cap, and commission breakeven."""
from __future__ import annotations
import threading
from datetime import date
from dataclasses import dataclass


@dataclass
class TradeRecord:
    symbol: str
    pnl: float
    timestamp: date


class RiskManager:
    def __init__(self, cfg: dict):
        self._cfg = cfg
        self._lock = threading.Lock()
        self._reset()

    def _reset(self):
        self._day = date.today()
        self._daily_pnl: float = 0.0
        self._trade_count: int = 0
        self._open_count: int = 0
        self._records: list[TradeRecord] = []

    def _check_day(self):
        if date.today() != self._day:
            self._reset()

    # --- public API ---

    def commission_for(self, trade_value: float) -> float:
        """One-way commission for a trade of the given SGD value."""
        rate = self._cfg["commission_rate"]
        min_c = self._cfg["min_commission_sgd"]
        return max(trade_value * rate, min_c)

    def can_open(self, trade_value_sgd: float) -> tuple[bool, str]:
        """Returns (allowed, reason). Call before placing an entry order."""
        with self._lock:
            self._check_day()
            max_pos    = self._cfg["max_position_sgd"]
            max_loss   = self._cfg["max_daily_loss_sgd"]
            max_open   = self._cfg["max_open_positions"]
            max_trades = self._cfg["max_trades_per_day"]

            if self._open_count >= max_open:
                return False, f"max open positions ({max_open}) reached"
            if self._trade_count >= max_trades:
                return False, f"max daily trades ({max_trades}) reached"
            # use a small buffer (0.01 SGD) so we catch the limit even with rounding
            if self._daily_pnl <= -(max_loss - 0.01):
                return False, f"daily loss limit reached (${max_loss:.2f} SGD)"
            if trade_value_sgd > max_pos:
                return False, f"position ${trade_value_sgd:.0f} > limit ${max_pos:.0f} SGD"
            return True, ""

    def expected_net_pnl(self, expected_gross: float, trade_value: float) -> float:
        """Net PnL after round-trip commission.
        expected_gross must already reflect direction (always positive for a profitable trade).
        """
        return expected_gross - 2 * self.commission_for(trade_value)

    def record_open(self):
        with self._lock:
            self._check_day()
            self._open_count += 1
            self._trade_count += 1

    def record_close(self, symbol: str, pnl: float):
        with self._lock:
            self._check_day()
            self._open_count = max(0, self._open_count - 1)
            # round each trade to avoid floating-point drift over many cycles
            self._daily_pnl = round(self._daily_pnl + pnl, 4)
            self._records.append(TradeRecord(symbol, pnl, date.today()))

    def summary(self) -> dict:
        with self._lock:
            self._check_day()
            return {
                "trades":    self._trade_count,
                "open":      self._open_count,
                "daily_pnl": round(self._daily_pnl, 2),
            }

File: trading-bot/internal/risk/manager.go
package risk

import (
	"fmt"
	"sync"
	"time"

	"github.com/mmedabo/trading-bot/config"
	"github.com/shopspring/decimal"
)

// TradeRecord tracks a completed round-trip.
type TradeRecord struct {
	Symbol    string
	PnL       decimal.Decimal
	Timestamp time.Time
}

// Manager enforces risk limits across the bot's lifecycle.
type Manager struct {
	mu           sync.Mutex
	cfg          *config.RiskConfig
	dailyPnL     decimal.Decimal
	tradesCount  int
	openCount    int
	records      []TradeRecord
	lastResetDay int // day-of-year of last reset
}

func New(cfg *config.RiskConfig) *Manager {
	return &Manager{cfg: cfg, lastResetDay: time.Now().YearDay()}
}

// resetIfNewDay clears daily counters when calendar day changes.
func (m *Manager) resetIfNewDay() {
	today := time.Now().YearDay()
	if today != m.lastResetDay {
		m.dailyPnL = decimal.Zero
		m.tradesCount = 0
		m.lastResetDay = today
	}
}

// CanOpen checks whether a new position may be opened for the given trade value (SGD).
func (m *Manager) CanOpen(tradeSGD decimal.Decimal) error {
	m.mu.Lock()
	defer m.mu.Unlock()
	m.resetIfNewDay()

	if m.openCount >= m.cfg.MaxOpenPositions {
		return fmt.Errorf("max open positions (%d) reached", m.cfg.MaxOpenPositions)
	}
	if m.tradesCount >= m.cfg.MaxTradesPerDay {
		return fmt.Errorf("max daily trades (%d) reached", m.cfg.MaxTradesPerDay)
	}
	if m.dailyPnL.LessThan(decimal.NewFromFloat(-m.cfg.MaxDailyLossSGD)) {
		return fmt.Errorf("daily loss limit reached (%.2f SGD)", m.cfg.MaxDailyLossSGD)
	}
	maxPos := decimal.NewFromFloat(m.cfg.MaxPositionSGD)
	if tradeSGD.GreaterThan(maxPos) {
		return fmt.Errorf("position size %s exceeds limit %s SGD", tradeSGD, maxPos)
	}
	return nil
}

// ExpectedPnL estimates net PnL after round-trip commissions.
// Returns an error if the trade won't be profitable.
func (m *Manager) ExpectedPnL(entryPrice, targetPrice decimal.Decimal, qty int64) (decimal.Decimal, error) {
	tradeValue := entryPrice.Mul(decimal.NewFromInt(qty))
	commission := commissionFor(tradeValue.InexactFloat64(), m.cfg.MinCommissionSGD, m.cfg.CommissionRate)
	roundTrip := commission.Mul(decimal.NewFromInt(2))

	gross := targetPrice.Sub(entryPrice).Mul(decimal.NewFromInt(qty))
	net := gross.Sub(roundTrip)

	if net.LessThanOrEqual(decimal.Zero) {
		return decimal.Zero, fmt.Errorf("trade not profitable after commission: gross=%.4f commission=%.2f",
			gross.InexactFloat64(), roundTrip.InexactFloat64())
	}
	return net, nil
}

// RecordOpen notes that a position was opened.
func (m *Manager) RecordOpen() {
	m.mu.Lock()
	defer m.mu.Unlock()
	m.openCount++
	m.tradesCount++
}

// RecordClose notes that a position closed with given PnL.
func (m *Manager) RecordClose(symbol string, pnl decimal.Decimal) {
	m.mu.Lock()
	defer m.mu.Unlock()
	m.openCount--
	if m.openCount < 0 {
		m.openCount = 0
	}
	m.dailyPnL = m.dailyPnL.Add(pnl)
	m.records = append(m.records, TradeRecord{
		Symbol:    symbol,
		PnL:       pnl,
		Timestamp: time.Now(),
	})
}

// DailyPnL returns total P&L for the current trading day.
func (m *Manager) DailyPnL() decimal.Decimal {
	m.mu.Lock()
	defer m.mu.Unlock()
	m.resetIfNewDay()
	return m.dailyPnL
}

// Summary returns a snapshot of today's stats.
func (m *Manager) Summary() (trades int, pnl decimal.Decimal, open int) {
	m.mu.Lock()
	defer m.mu.Unlock()
	m.resetIfNewDay()
	return m.tradesCount, m.dailyPnL, m.openCount
}

func commissionFor(value, minCommission, rate float64) decimal.Decimal {
	calc := value * rate
	if calc < minCommission {
		calc = minCommission
	}
	return decimal.NewFromFloat(calc).Round(2)
}
